### DISCUSSION TITLE
✋Looking for maintainers

### DISCUSSION THREAD

**Discussion by mweststrate** [mobxjs/mobx#1983] [General]
I am looking for maintainers of this repository. That probably won't be too much of a burden (no big features in the planning), as it is mostly reviewing an occasional PR and releasing it. But there are some funds available to compensate if it is gonna cost you too much time.

[MobX development is NOT terminated](https://github.com/mobxjs/mobx/issues/1983#issuecomment-593342914)

[How to start contributing](https://github.com/mobxjs/mobx/issues/1983#issuecomment-565981010)

**Comment by ItamarShDev** [comment]
I will be happy to help
not sure I am qualified though

**Comment by mweststrate** [comment]
@ItamarShDev no worries, I can try to guide you through it the first few times! And I think @capaj and @urugator will be willing to chime in with their knowledge were needed as well. It is mostly making sure things don't linger around for too long!

For my own sanity, I am going to unwatch the repositories, but if my input is needed, just nudge!

**Comment by ItamarShDev** [comment]
Cool!
Well if any of the members will agree, I will be happy to maintain and help as much as i can!

Also, I hope i could get some guidelines and guidance, so i could do the best job i can :)

**Comment by mweststrate** [comment]
I'll compile a list in the coming days! Ping me if I didn't :)

**Comment by spion** [comment]
cc @benjamingr

**Comment by benjamingr** [comment]
Hey, I don't mind picking up more of the burden here.

We can also adopt this at the open source initiative in jsil.

Also - consider joining the OpenJS foundation 

Feel free to ping me about any or all of this in private channels :)

**Comment by aretecode** [comment]
feel free tag or request me for any reviews 👍

**Comment by StephenHaney** [comment]
Time is always tight, but I’d love to support however possible! Big fan of the project.

**Comment by mweststrate** [comment]
Awesome guys, thanks a lot! I'll try to compile a list of things I
generally do in a couple of days!

**Comment by danielkcz** [comment]
@mweststrate Are you going to give access for NPM publish to everyone as well? I am not saying people are generally trying to do any harm, but more people with access means potential security hole. I think it would be wise to require 2FA enabled for anyone who will be pushing to NPM.

**Comment by rdewolff** [comment]
Would be happy to help as well, with some guidance.

**Comment by mweststrate** [comment]
Thanks guys for the responses! Heartwarming! Will send invites in an iffy. 

@FredyC I think we can keep NPM access currently limited to you, xavier, itamar and those that already have it? I have no problems enrolling more people after being active in the repository for a while.

**Comment by mweststrate** [comment]
Maintainer tips

Hi folks!

Thanks for the interest in maintaining MobX! Here are some practical tips and general experiences that might help to become an effective maintainer. It is not all black and white, these are just things that worked for me, but if someone else works for you, or seems better for the project, just feel free to deviate!

## General thoughts

1. Not all problems require solving; many questions are quite conceptual. And it might be valuable to answer them, but it is not mandatory. Most devs are paid in the end to work their own way through a problem. Many vaguely formulated problems have their solution in either more strictly adhering to the core principles of MobX and the separation of state, effects, etc. And, on the other hand, there are edges that don't fit nicely, or don't perform nicely in the standard model of MobX. In that case people should feel comfortable in using the provided escape hatches. That is what escape hatches are for (including: not using MobX for a certain feature at all). 
2. It is primarily the responsibility of issue filers to isolate and reproduce issues. Especially on a code base that is as mature and widely used as MobX. 
3. That doesn't mean that MobX is bug free, but it doesn't mean either that every bug needs to be fixed. Some bugs are beyond our control (browsers), some are deliberately left there; if the edge case is too uncommon, I rather avoid having the whole community pay for a slower MobX to address the edge case of one user. This is always a balancing act. But the product doesn't have to be perfect. It should work in the expected cases and have the limitations documented. Typical case: observable on statics is not supported, even though it might work, it is not something we guarentee
4. Be prepared to take a lot of initiative as maintainer; there will be a lot of input in terms of issues and PRs, but nobody to boss you around to decide what should be solved. Always keep the interest of the community at large in mind with whatever you pick up (in terms of api surface size, bundle size, performance, backward compatibility and consistency accross different versions). Always feel free to pull in the expertise of other active community members by pinging them (in my case, if that doesn't work, PM me on twitter)
5. Answering recurring questions in the docs really pays of. I used to do that for a long time and it really brought down the amount of issues raised (or they were much more easily closed). However, the docs around 'best practices' are still pretty limited.

## Practical tips

* Both new features and bug fixes should always be covered by tests
* Releasing often makes it easier. Don't worry too much about accidentally breaking stuff, reverting is easy, everybody uses lock files nowadays, and breaking something in the wild is a good teacher. 
* In principle every feature and bug fix should be merged back to mobx4-master (I usually cherry-pick relevant commits for that). Exceptions are things that relate to Proxies and Symbols. 
* The changelog is only maintained on master, for both mobx 4 and 5. Currently that is a manual process.

## The release process 

1. Review PR. Pay special notice to
   - do the docs need an update? In that case request a PR on the `gh-pages` branch
   - should it be backported to MobX 4? You can request a pr for that, or do it yourself. 
   - are there sufficient tests?
1. Merge the PR to master. Usually I use squash (if there are conflicts you might ask to rebase, or solve conflicts yourself)
5. Push, check if CI succeeds
3. Update `changelog.md`, make sure to mention issues solved, the PR, the author
4. Run `./scripts/publish.js` to bump the version number and publish to NPM (don't do this by hand!)
5. Go to the `mobx4-master` branch and cherry-pick the relevant commits
5. Push, check if CI succeeds
6. Run `./scripts/publish.js` (this will automatically make sure the right npm tag is used for MobX 4!)

To republish the docs:
1. Checkout the `gh-pages` branch
2. Switch to Node 6.10 (using nvm) (sorry, the docs setup is really ancient and horrible)
3. Run `./build.sh`
4. Commit all and push


## Funds

There are some limited funds available, provided by our backers on https://opencollective.com/mobx. If you are spending significant own time on MobX on a regular basis, feel free to reach out!

## ... and many things more

There are probably many things missing here, I'll try to add them when they pop up in my mind!

**Comment by danielkcz** [comment]
@mweststrate Thanks for the write-up, very useful and insightful indeed.

> I think we can keep NPM access currently limited to you, xavier, itamar and those that already have it? I have no problems enrolling more people after being active in the repository for a while.

What I meant is that NPM is not securely tied up to GitHub code. There has been already reported cases of stolen NPM accounts and the culprit can then easily publish a version with some extra malware code. It's very hard to track that down.

**Comment by benjamingr** [comment]
@lirantal any idea how we can do this more securely?

Honestly I can already publish very scary packages with way more downloads than MobX but I would never do that even if I was a crook because my livelihood depends on it and I'd get (sued) and am reasonably easy to locate.

As for Itamar, I know where @ItamarShDev lives and where his family lives so there's that 😃 More seriously his livelihood is also related to his web development career and his employer is decent accountability. 

I don't know Xaviar well enough to say anything about him being able to publish things :] 

Moreover publishing is a pain and if we can get our CI to do it (and no one else) from GitHub releases and not give anyone (but Michel) the keys that would be my most preferred option.

**Comment by mweststrate** [comment]
I guess we will have to wait a little bit until Microsoft finally acquires
NPM and solves this problem fundamentally :-P

**Comment by danielkcz** [comment]
> Moreover publishing is a pain and if we can get our CI to do it (and no one else) from GitHub releases and not give anyone (but Michel) the keys that would be my most preferred option.

That sounds like a good way to go. Less human aspect involved could be better indeed. Do you have some experience setting that up?

**Comment by benjamingr** [comment]
@FredyC I do, I don't however have time in the coming month or two unfortunately :( I don't mind guiding other people in doing that it's basically:

 - Open an account on CircleCI (this is free)
 - Use a generated config.yml
 - Run npm 

Here is a tutorial on how it's typically done it's pretty straightforward https://circleci.com/blog/publishing-npm-packages-using-circleci-2-0/ 

It would involve (hopefully):

 - Michel creating a new account and giving it a token just to publish this package (rather than his full account)
 - Michel copying that token to circle
 - Michel is then the only person with access to the CI to update the token, everyone else just pushes to GitHub and draft releases.

**Comment by ItamarShDev** [comment]
@mweststrate Thank you  very much for the long and thorough explanation!
I will do my best to stay active in all of MobX repos (but mst) and help as much as I can.    

As for releasing and deciding which bugs should not be fixed, I will see how other members decide and will start involving when I will fill more comfortable in this area.

@benjamingr I will gladly help setting up a CCI publisher if needed

@mweststrate  is there any benefit from upgrading the `gh-pages` to be on the same node version?
We can suggest it to new contributors that are looking for a way into OSS, if it's not hard

**Comment by lirantal** [comment]
> @lirantal any idea how we can do this more securely?

The npm security cheatsheet I published has some good tips on developer's own security best practices that should be adopted so I suggest reading through that. Might be relevant - I'm working with some other open source projects on a new program I'm piloting about establishing security practices for the project, starting from devs, to CI, and such. If @mweststrate and the new dev team here is interested to follow-up on that I'm happy to work with you fellas on this.

> Moreover publishing is a pain and if we can get our CI to do it (and no one else) from GitHub releases and not give anyone (but Michel) the keys that would be my most preferred option.

>> That sounds like a good way to go. Less human aspect involved could be better indeed. Do you have some experience setting that up?

This is not inherently more secure since someone can backdoor a contribution in a PR that got through and it would get automatically (and very quickly) released without further humans watching it.

That said, I agree with the need for automation and the plan that @benjamingr set out in https://github.com/mobxjs/mobx/issues/1983#issuecomment-500765458 makes sense to follow.

**Comment by danielkcz** [comment]
> This is not inherently more secure since someone can backdoor a contribution in a PR that got through and it would get automatically (and very quickly) released without further humans watching it.

That's true, but at least you have some sort of evidence in the source code that can be tracked and reverted. In case someone hacks into an NPM account, they can publish whatever they want. Sure we can always publish over with "correct version", but it's way more sneaky in my opinion.

I think it would be nice to set up the exact same workflow for all MobX repos which @mweststrate does not want to publish by hand anymore. That includes current `mobx-react-lite` where only I have access for publishing. It's would be much easier to maintain such an ecosystem with the same tooling all over the place without thinking "ah, it works this way in this repo".

I am also thinking about employing [semantic-release](https://github.com/semantic-release/semantic-release) to remove another level of human intervention. I love how it automatically makes a release after merging PR and sends a message to the PR. It works great in eg. [react-testing-library repo](https://github.com/testing-library/react-testing-library/pull/365#issuecomment-486841590). That way maintainers can fully focus on the quality of the PR and forget about any tooling.

**Comment by mweststrate** [comment]
Immer is being released by semantic- release atm, and it is really great
indeed. I think we can follow a similar setup here.

Op wo 12 jun. 2019 12:23 schreef Daniel K. <notifications@github.com>:

**Comment by danielkcz** [comment]
We should probably figure out automatic releases soon. It's getting somewhat harder to reach someone who can publish and fixes are just lying around without people being able to use them. It's definitely something annoying to do hence automatic process makes total sense,

From my understanding of summary in https://github.com/mobxjs/mobx/issues/1983#issuecomment-500765458 it mostly depends on @mweststrate to set it up. Are there some other steps we can make the process easier? 

I also would like to employ it for other repos so we have a crystal clear view into publishing process across the MobX community.

https://github.com/mobxjs/mobx-react
https://github.com/mobxjs/mobx-react-lite
https://github.com/mobxjs/mobx-state-tree
https://github.com/mobxjs/mobx-utils

**Comment by mweststrate** [comment]
@FredyC I've just granted you access to the above repositories on NPM. 

The release process is now pretty uniform:

1. do all the merging (and wait for succesful master build)
1. update changelog.md (this is the only part that consumes a little time)
2. run `./publish.js`

With immer, we automated this indeed with semantic-release. It is quite great, but one down side is that the commit messages need to be standardized, which is potentially an additional burden for new contributors (as they need to adhere to the standard), or we should squash all the merges. So that is why I never adopted it so far. But in itself it works really great, so I don't have any objections as long as we make sure new contributions are still streamlined. 

Afaik everybody can set that up in circle / travis, doesn't necessary have to be me (@aleclarson did it for immer, now all release nicely bear his name :grinning: )

**Comment by danielkcz** [comment]
> which is potentially an additional burden for new contributors (as they need to adhere to the standard)

Contributors don't need to actually worry about that. When squashing PR the maintainer can set different commit message to express what kind of change is that. 

Thanks for permissions, I will try to figure this out soon.

**Comment by mweststrate** [comment]
So far we didn't do much squashing. In general I'm pro squashing PRs, the
only downside of squashing in MobX specifically is that it makes it harder
to cherry-pick specific commits from mobx 5 to mobx 4.

**Comment by danielkcz** [comment]
Well, technically you can still cherry-pick commits from that branch, squashing does not remove them from Git. And I would dare to say that PR is usually so narrowly focused that you want to grab a whole thing. Point is that maintainers can shield away nuances of `semantic-release`.

I would probably try out first those automated GH releases without semantic release, that alone will surely help current situation.

**Comment by mweststrate** [comment]
Doh, good point.

Feel free to play with it. Let me know if you get stuck somewhere.
Especially getting the gh release working in the first place might be a
problem. If you can't get it to work locally, I'll try to make another
clone locally, and check what needs to be done to get it similar to my
'original' checkout.

**Comment by Jongkeun** [comment]
Wow @mweststrate 👏
 I was just one of the user.
I was interested in this project.
So I thought I would try to contribute to this project and I hoped to be a maintainer of this project.
Can I be a maintainer too?
Are there any requirements?

**Comment by danielkcz** [comment]
@Jongkeun What do you actually expect? :) Just be active in a community, respond to user issues, do code review, provide own PR for open issues or be helpful however you can. You don't need any special title for any of that.

**Comment by Jongkeun** [comment]
@FredyC 
Haha, I expected mark of Organizations. 😁
Once, I think, being active in this community is the first step as you said. I will try to do with pleasure.

**Comment by danielkcz** [comment]
Just leaving this here as it looks somewhat interesting.

https://help.github.com/en/github/managing-packages-with-github-packages/about-github-packages

I know I wanted to tackle the publishing process with semantic-release and stuff but just cannot get through to it. Things happening all the time lately :/ Also, GitHub Actions sounds an interesting approach.

**Comment by auvipy** [comment]
I would love to be a co-maintainer :1st_place_medal:

**Comment by danielkcz** [comment]
@auvipy https://github.com/mobxjs/mobx/issues/1983#issuecomment-529821469 or https://github.com/mobxjs/mobx-react/issues/688#issuecomment-558197583

**Comment by tzachbon** [comment]
Hi, I would love to be help and be a maintainer!

**Comment by danielkcz** [comment]
Come on, folks, you want to be a maintainers and you don't even read the last comment? I don't want to be repeating it for everyone... :/

**Comment by mweststrate** [comment]
Hi @auvipy @tzachbon thanks and welcome! 

So far becoming maintainer has always been an organic process, and I think it is a good one. So just chime in on the issues, review PRs, create fixes or propose things. Generally we grant out further permissions on a need basis; if you're active in discussions, we make sure you get moderator permissions asap. If you'r active in the code base, you'll get repo access asap, so that you don't have to run your own fork, etc etc! 

So just chime in and we'll do our best to guide you along the way! (In practice, that is honestly less than it should be, but we'll try ;-)). Some things on which help is always welcome:

* Review / follow up on PR progress
* Triage issues, make sure the proper information is surfaced, and that below par issues are closed
* Submit PR for things that are marked with the `bug` labels

I don't recommend to start with proposing or creating big chances; in generally we are pretty conservative in making big changes for two reasons 1) for thousands of developers things are working as they are, and we don't want to introduce very subtle changes for them 2) we want to make sure that you're in it for the long run; if someone introduces a big, very good change, but isn't there to support it, it in the end often hurts more than it helps as it forces maintenance burden on others (trust me, speaking from experience here). 

If you have no idea what to pick up first, here are two suggestions:
1. There are a lot of open questions on mobx-state-tree, if you have experience with it, chime in! It is currently an underinvested area
2. Mobx-react-lite 2.0 is practically finished, but awaiting more testing. If you have an opportunity to put mobx-react-lite (or a mobx-react build with lite 2.0) in a real project and confirm it is working as expected, that would be really valuable!

Thanks!

**Comment by auvipy** [comment]
sure! i will! i use mobx on pro projects with react

**Comment by rumax** [comment]
> I am looking for maintainers of this repository. That probably won't be too much of a burden (no big features in the planning), as it is mostly reviewing an occasional PR and releasing it. But there are some funds available to compensate if it is gonna cost you too much time.

Hi @mweststrate, would you be so kind to elaborate a bit more about the reason of such decision? How this will affect new features/releases, etc? 

Is it still good idea to use `mobx` in the existing projects **or we should think about migration to other solutions**? 

In case of a new project, what are your expectations for `mobx` as a project after it is transferred to new maintainers? Will it still have significant improvements, new versions or it will be frozen at the current state with only some bug fixes?

**Comment by danielkcz** [comment]
@rumax Don't worry, MobX isn't going anywhere, there are other maintainers. Michel has several other projects which he needs to split attention to, but he is still very active and we are even planning some major changes (see #2288)

**Comment by stolenng** [comment]
I would like to start contributing, what is the process of joining? :)

**Comment by danielkcz** [comment]
@stolenng Please see https://github.com/mobxjs/mobx/issues/1983#issuecomment-565981010

**Comment by lukigarazus** [comment]
Is this still a possibility?

**Comment by danielkcz** [comment]
@lukigarazus There is always some work around here, tending to other people questions/issues at least. I recommend you to read https://github.com/mobxjs/mobx/issues/1983#issuecomment-565981010

**Comment by Rain120** [comment]
I will try it in my free time, now I had tried to understand the code at the version of mobx 6. Fighting 💪 💪 💪

**Comment by havit1** [comment]
Is this still relevant? And if so, is this relevant for someone with no open sourse experience?

**Comment by dante01yoon** [comment]
Any open position on this? I'm interested in

**Comment by kubk** [comment]
@dante01yoon @havit1 You don't need to ask permission to be useful for the project. Please read: https://github.com/mobxjs/mobx/discussions/1983#discussioncomment-104567

**Reply by dante01yoon** [reply]
I see. Thank you

**Reply by mweststrate** [reply]
@dante01yoon welcome and feel free to chime in dante!
